Add TabIndex label to the statusbar.

ff2000 opened the pull request:
It shows the current tab index and the number of tabs of the window it
sits in.

The-Compiler commented:
Thanks a lot for your contribution!

I think it'd be better to not access the tabwidget from `mainwindow/statusbar/tabindex.py` however, so it's more self-contained. What about this?

* Add a new `tab_index_changed = pyqtSignal(int, int)` to `TabWidget` in `qutebrowser/mainwindow/tabwidget.py`
* Connect `QTabWidget::currentChanged` to a new slot in `TabWidget`, where you emit the `tab_index_changed` signal with `idx + 1` and `self.count()` as arguments.
* Then connect that new signal to your `on_tab_changed` and use the values you get there.

I'll also leave some inline comments with some minor nitpicks.

The-Compiler commented on qutebrowser/mainwindow/statusbar/tabindex.py:
```diff
@@ -0,0 +1,42 @@
+    @pyqtSlot(int)
+    def on_tab_changed(self, tab):
+        """Update tab index when tab changed."""
+        self.setText('[{0}/{1}]'.format(
```
Please simply use `[{}/{}]` here - python3 doesn't need the explicit numbering :)

ff2000 replied:
Florian Bruhin wrote:
> > +from qutebrowser.utils import objreg
> > +
> > +class TabIndex(textbase.TextBase):
> > +
> > +    """Shows current tab index and number of tabs in the statusbar."""
> > +
> > +    def __init__(self, win_id, parent=None):
> > +        super().__init__(parent)
> > +        self.tabbed_browser = objreg.get('tabbed-browser',
> > +                                         scope = 'window',
> > +                                         window = win_id)
> > +
> > +    @pyqtSlot(int)
> > +    def on_tab_changed(self, tab):
> > +        """Update tab index when tab changed."""
> > +        self.setText('[{0}/{1}]'.format(
> 
> Please simply use `[{}/{}]` here - python3 doesn't need the explicit numbering :)
ARGH! :)
I read the docs about str.format() and saw "2.7 only". Had another look and it actually says "2.7+ only". YAY :)
(I am not a native pythonic, so please forgive me...)

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/The-Compiler/qutebrowser/pull/596/files#r27636851

The-Compiler commented on qutebrowser/mainwindow/statusbar/tabindex.py:
```diff
@@ -0,0 +1,42 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2015 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
```
That should be 2015 only, and you might prefer your own name/mail here - but if you prefer, mine is okay as well.

ff2000 commented:
Florian Bruhin wrote:
> Thanks a lot for your contribution!
> 
> I think it'd be better to not access the tabwidget from `mainwindow/statusbar/tabindex.py` however, so it's more self-contained. What about this?
> 
> * Add a new `tab_index_changed = pyqtSignal(int, int)` to `TabWidget` in `qutebrowser/mainwindow/tabwidget.py`
> * Connect `QTabWidget::currentChanged` to a new slot in `TabWidget`, where you emit the `tab_index_changed` signal with `idx + 1` and `self.count()` as arguments.
> * Then connect that new signal to your `on_tab_changed` and use the values you get there.
That actually was my first thought, too. Decoupling types usually is the better approach.
I decided against it because TabIndex would have been the only consumer of the signal. But I don't have any strong argument against it, so I will go and implement it that way.

> 
> I'll also leave some inline comments with some minor nitpicks.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/The-Compiler/qutebrowser/pull/596#issuecomment-88788041

ff2000 commented:
I have issues with connections in TabWidget :(

```
diff --git a/qutebrowser/mainwindow/tabwidget.py b/qutebrowser/mainwindow/tabwidget.py
index 590f0e8..6e82723 100644
--- a/qutebrowser/mainwindow/tabwidget.py
+++ b/qutebrowser/mainwindow/tabwidget.py
@@ -26,7 +26,7 @@ Module attributes:
 
 import functools
 
-from PyQt5.QtCore import pyqtSlot, Qt, QSize, QRect, QPoint, QTimer
+from PyQt5.QtCore import pyqtSignal, pyqtSlot, Qt, QSize, QRect, QPoint, QTimer
 from PyQt5.QtWidgets import (QTabWidget, QTabBar, QSizePolicy, QCommonStyle,
                              QStyle, QStylePainter, QStyleOptionTab)
 from PyQt5.QtGui import QIcon, QPalette, QColor
@@ -43,6 +43,8 @@ class TabWidget(QTabWidget):
 
     """The tab widget used for TabbedBrowser."""
 
+    tab_index_changed = pyqtSignal(int, int)
+
     def __init__(self, win_id, parent=None):
         super().__init__(parent)
         bar = TabBar(win_id)
@@ -50,6 +52,7 @@ class TabWidget(QTabWidget):
         bar.tabCloseRequested.connect(self.tabCloseRequested)
         bar.tabMoved.connect(functools.partial(
             QTimer.singleShot, 0, self.update_tab_titles))
+        bar.currentChanged.connect(self.on_current_changed)
         self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Fixed)
         self.setDocumentMode(True)
         self.setElideMode(Qt.ElideRight)
@@ -58,6 +61,11 @@ class TabWidget(QTabWidget):
         self.init_config()
         objreg.get('config').changed.connect(self.init_config)
 
+    @pyqtSlot(int)
+    def on_current_changed(self, index):
+        print("tab index changed {}{}".format(index, self.count))
+        self.tab_index_changed.emit(index, self.count())
+
     @config.change_filter('tabs')
     def init_config(self):
         """Initialize attributes based on the config."""
```
The new SLOT is never called. It also doesn't work if I use self.current Changed instead of bar.currentChanged.
If I add the new signal directly to TabbedBrowser everything just works as expected. There is already a SLOT connected to currentChanged, so it's just two extra lines. If I do this should I change the signal to cur_tab_index_changed to match naming)?

Is this caused by Signal filtering?

The-Compiler commented:
There we go - thanks for the fast changes! :)

pylint still complained because the import line was too long, but I changed that myself :)

ff2000 commented:
Florian Bruhin wrote:
> There we go - thanks for the fast changes! :)
Thx for the fast review :)
qutebrowser finally seems to be what I've been looking for:
* much faster interface than firefox
* vim-like interface from the beginning (would be hard to integrate into qupzilla/otter)
* more sane codebase than e.g. uzbl
* *native* tabbing - no hacks like tabbed (the app)
(last point will be the target of my next effort: implement ":buffer" to get rid of the tabbar)

And I reallyreallyreally hope QtWebKit will stay stable now... The last times I tried to use QtWebKit it instantaniously started crashing :(

> 
> pylint still complained because the import line was too long, but I changed that myself :)
Is there some sort of vim-integration available?
* settings for the editor (max columns, indentation, ...)
* syntastic integration for everything else?

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/The-Compiler/qutebrowser/pull/596#issuecomment-88854188

The-Compiler commented:
> * *native* tabbing - no hacks like tabbed (the app)

Hah - finally someone who agrees with me! I don't think "tabbing doesn't belong in a browser, that's not the Unix way!!1!" is a sane thing to say :D

> (last point will be the target of my next effort: implement ":buffer" to get rid of the tabbar)

Cool! Also check #32 for a related issue.

> And I reallyreallyreally hope QtWebKit will stay stable now... The last times I tried to use QtWebKit it instantaniously started crashing :(

I've reported [many QtWebKit bugs](https://bugreports.qt.io/browse/QTBUG-45300?jql=component%20%3D%20WebKit%20and%20reporter%20%3D%20%22Florian%20Bruhin%22%20and%20priority%20in%20(%22P0%3A%20Blocker%22%2C%20%22P1%3A%20Critical%22%2C%20%22P2%3A%20Important%22)) and most of them were promptly fixed, i.e. things are definitely getting better and better.

Also at some point qutebrowser will add a QtWebEngine backend which uses Blink/Chromium - that's probably more stable and modern.

> > pylint still complained because the import line was too long, but I changed that myself :)
> Is there some sort of vim-integration available?
> * settings for the editor (max columns, indentation, ...)
> * syntastic integration for everything else?

The vim modelines should take care of the line width - though some things probably depend on your config too.

The last time I looked at syntastic it was really buggy so I'm not using it, maybe I should take another look at it!

I recommend you install [tox](https://tox.readthedocs.org/en/latest/) - that's what qutebrowser uses to run all tests and linters. Simply run `tox` before commiting, wait a while, and hope everything is green ;)

Ram-Z commented:
> pylint still complained because the import line was too long, but I changed that myself :)
> Is there some sort of vim-integration available?
> * settings for the editor (max columns, indentation, ...)
> * syntastic integration for everything else?

There's [python-mode](https://github.com/klen/python-mode), it's a bit on the heavy side though. I recommend using [wilywampa's fork](https://github.com/wilywampa/python-mode) though, it has a few fixes already implemented.

Syntastic has pylint support, if it can find the executable in your `$PATH`. ([Syntastic FAQ](https://github.com/scrooloose/syntastic#4-faq))
